File: forms/medical-error-report/front-end-form-with-html/js/error-grader.js
// Medical-error grader. Pure functions: take an `AssessmentData` object,
// evaluate every rule, derive WHO severity, NCC MERP category, and overall
// risk level. Mirrors the SvelteKit `error-grader.ts`.

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').FiredRule} FiredRule
 * @typedef {import('./types.js').GradingResult} GradingResult
 * @typedef {import('./types.js').WHOSeverity} WHOSeverity
 * @typedef {import('./types.js').NCCMERPCategory} NCCMERPCategory
 * @typedef {import('./types.js').RiskLevel} RiskLevel
 */

(function () {
'use strict';
window.MedicalErrorReport = window.MedicalErrorReport || {};
const { errorRules, detectAdditionalFlags } = window.MedicalErrorReport;

/**
 * Evaluate every error rule, then derive WHO severity, NCC MERP category,
 * and overall risk level.
 *
 * @param {AssessmentData} data
 * @returns {GradingResult}
 */
function calculateErrorGrade(data) {
  /** @type {FiredRule[]} */
  const firedRules = [];

  for (const rule of errorRules) {
    try {
      if (rule.evaluate(data)) {
        firedRules.push({
          id: rule.id,
          category: rule.category,
          description: rule.description,
          grade: rule.grade
        });
      }
    } catch (e) {
      console.warn(`Error rule ${rule.id} evaluation failed:`, e);
    }
  }

  const whoSeverity = data.errorClassification.whoSeverity || '';
  const nccMerpCategory = data.errorClassification.nccMerpCategory || '';
  const overallRisk = deriveOverallRisk(firedRules, whoSeverity, nccMerpCategory);
  const additionalFlags = detectAdditionalFlags(data);

  return {
    whoSeverity,
    nccMerpCategory,
    overallRisk,
    firedRules,
    additionalFlags,
    timestamp: new Date().toISOString()
  };
}

/**
 * Derive overall risk from worst fired-rule grade and reporter classifications.
 *
 * @param {FiredRule[]} firedRules
 * @param {WHOSeverity} whoSeverity
 * @param {NCCMERPCategory} nccMerpCategory
 * @returns {RiskLevel}
 */
function deriveOverallRisk(firedRules, whoSeverity, nccMerpCategory) {
  const maxGrade =
    firedRules.length > 0 ? Math.max.apply(null, firedRules.map((r) => r.grade)) : 0;

  const whoRisk = {
    'near-miss': 1,
    'mild': 2,
    'moderate': 3,
    'severe': 4,
    'critical': 5
  };
  const merpRisk = {
    A: 1, B: 1, C: 1,
    D: 2,
    E: 3, F: 3,
    G: 4,
    H: 5, I: 5
  };

  const whoLevel = whoSeverity ? (whoRisk[whoSeverity] || 0) : 0;
  const merpLevel = nccMerpCategory ? (merpRisk[nccMerpCategory] || 0) : 0;
  const effectiveMax = Math.max(maxGrade, whoLevel, merpLevel);

  if (effectiveMax >= 5) return 'critical';
  if (effectiveMax >= 4) return 'high';
  if (effectiveMax >= 3) return 'moderate';
  return 'low';
}

Object.assign(window.MedicalErrorReport, { calculateErrorGrade });
})();

File: forms/medical-error-report/full-stack-with-loco-tera-htmx-alpine/src/engine/error_grader.rs
use super::error_rules::all_rules;
use super::flagged_issues::detect_additional_flags;
use super::types::{AssessmentData, FiredRule, GradingResult, RiskLevel};

/// Pure function: grades a medical-error-report case.
///
/// Returns a [`GradingResult`] containing:
/// - WHO severity (from reporter input)
/// - NCC MERP category (from reporter input)
/// - overall risk level (derived from worst fired-rule grade + WHO + MERP)
/// - fired grading rules
/// - additional safety / governance flags
pub fn calculate_error_grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let fired_rules = run_rules(data);
    let additional_flags = detect_additional_flags(data);

    let who_severity = data.error_classification.who_severity.clone();
    let ncc_merp_category = data.error_classification.ncc_merp_category.clone();
    let overall_risk = derive_overall_risk(&fired_rules, &who_severity, &ncc_merp_category);

    GradingResult {
        who_severity,
        ncc_merp_category,
        overall_risk,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

/// Evaluate every rule against the data, returning the rules that fired.
pub fn run_rules(data: &AssessmentData) -> Vec<FiredRule> {
    let mut out = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data) {
            out.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                grade: rule.grade,
            });
        }
    }
    out
}

/// Derive overall risk level from the worst fired-rule grade combined with
/// reporter-supplied WHO severity and NCC MERP category.
fn derive_overall_risk(
    fired_rules: &[FiredRule],
    who_severity: &str,
    ncc_merp_category: &str,
) -> RiskLevel {
    let max_grade = fired_rules.iter().map(|r| r.grade).max().unwrap_or(0);

    let who_level = match who_severity {
        "near-miss" => 1,
        "mild" => 2,
        "moderate" => 3,
        "severe" => 4,
        "critical" => 5,
        _ => 0,
    };

    let merp_level = match ncc_merp_category {
        "A" | "B" | "C" => 1,
        "D" => 2,
        "E" | "F" => 3,
        "G" => 4,
        "H" | "I" => 5,
        _ => 0,
    };

    let effective_max = max_grade.max(who_level).max(merp_level);

    if effective_max >= 5 {
        "critical".to_string()
    } else if effective_max >= 4 {
        "high".to_string()
    } else if effective_max >= 3 {
        "moderate".to_string()
    } else {
        "low".to_string()
    }
}
